### DISCUSSION TITLE
Preferred way to introduce a boolean member value?

### DISCUSSION THREAD

**Discussion by kfsone** [badicsalex/peginator#26] [Q&A] [answered]
For me this often occurs with arity designation ("this thing is an array type"), like go's `[]string` (reads: array of strings).

Seems I could probably solve this with a custom parser, but I want to allow the full gamut of

```
[]string
[
]string
[/*who knows how many there are*/] string
```

I could just cheeze it and make `arity` be a string, and test for presence/emptiness.

```
@string
IsArray = @:ArityDesignation;

@string
ArityDesignation = OpenBrace CloseBrace;

@char
OpenBrace = '[';
@char
CloseBrace = ']';
```

but I was hoping there was some way to sneak in a literal of my own

```
@string
IsArray = "[" "]" is_array:<true> | !"[" is_array:<false>;
```

?

**Comment by badicsalex** [comment, marked as answer]
I'm not sure if I understand the example correctly, but what you could probably do is

```
Variable = [is_array:ArrayMarker] name:Identifier;

ArrayMarker = "[" "]";

@string
@no_skip_ws
Identifier = {IdentifierChar}+;

@char
IdentifierChar = 'a'..'z' | 'A'..'Z' | '0'..'9' | '_';
```

And then check for `Some`-ness of `variable.is_array`.

**Reply by kfsone** [reply]
That's more-or-less what I assumed the solution would be, what I was hoping for was a way to reduce it down to a bool (that's a stretch for an obscure edge case like this) or have the production return a constant value;

```
ArrayMarker: bool = "[" "]";
```

such that in the rule using this, say

```
MemberDecl = name:Identier [is_array:ArrayMarker] type:Identifier ',';
```

then MemberDecl.is_array is just a bool.

**Reply by badicsalex** [reply]
This is not the first time this topic came up: https://github.com/badicsalex/peginator/issues/18#issuecomment-1690707610

To do this, while expanding MemberDecl, the code would probably need to look at ArrayMarker and see that it is actually an empty type, and special-case the member type based on this info.
Unfortunately rules are expanded into types and parsers (mostly) independently, so this would need some refactoring to do (e.g. a preprocessing step to get a set of empty types)

What I really wouldn't want is manual marking (either something like `@marker` at rule level, or a special syntax at field level).
